cpp/src/arrow/filesystem/filesystem_test.cc

```diff
@@ -265,6 +265,24 @@ TEST(PathUtil, ToSlashes) {
 #endif
 }
 
+TEST(PathUtil, Globber) {
+  Globber localfs_linux("/f?o/bar/a?/1*.txt");
+  ASSERT_TRUE(localfs_linux.Matches("/foo/bar/a1/1.txt"));
+  ASSERT_TRUE(localfs_linux.Matches("/f#o/bar/ab/1000.txt"));
```

Comment:
Can you test failing matches as well?
For example, what happens when matching `"/foo/bar/ab/1/23.txt"`?

Reply:
Well, this path should be accepted because * includes /, but in practice '/' is not allowed to be in a filename, I'll add something to avoid this, and I'll add that test as a failing one.

Reply:
Substrait does not seem to specify precisely how a glob works, but I don't think `*` should match the path separator.

Reply:
Substrait does specify POSIX syntax and [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/) specifies:

```
The <slash> character in a pathname shall be explicitly matched by using one or more <slash> characters in the pattern; it shall neither be matched by the <asterisk> or <question-mark> special characters nor by a bracket expression.
```
So `*` should not match the path separator.